### PR TITLE
[TypeScript] More specific core-data selector types

### PR DESCRIPTION
**Pull request by adamziel**
## What?

This PR narrows down `core-data` selectors type signatures after somewhat general ones were introduced as a first step in https://github.com/WordPress/gutenberg/pull/41235.

## Why?

We want to have autocompletion and TypeScript support for core-data primitives as well as compile-time checks, see https://github.com/WordPress/gutenberg/pull/39025 for more context.

## Testing Instructions

* Confirm the CI checks are green
* Confirm the few runtime changes this PR introduces to make TS happy are sound

@noisysocks @sirreal @dmsnell @sarayourfriend

### CODE REVIEW

File: docs/reference-guides/data/data-core.md [hunk 1]
```diff
@@ -39,9 +39,9 @@ Calling this may trigger an OPTIONS request to the REST API via the
 _Parameters_
 
 -   _state_ `State`: Data state.
--   _kind_ `Kind`: Entity kind.
--   _name_ `Name`: Entity name.
--   _recordId_ `GenericRecordKey`: Record's id.
+-   _kind_ `K`: Entity kind.
+-   _name_ `N`: Entity name.
```
Comment (dmsnell) [inline docs/reference-guides/data/data-core.md]: when I mentioned having a different name for the public vs. the internal use of function arguments this is one of the reasons why. we like to have short names internally but `K` and `N` are vague to the outside.

we can solve this in the source files by importing `* as …`

```ts
import { …normalImports } from './entity-types';
import type * as ET from './entity.types';

export const doer<Kind extends ET.Kind, Name extends ET.Name>(…) => …
```

Comment (adamziel) [reply docs/reference-guides/data/data-core.md]: Good point! I just updated this PR @dmsnell

File: docs/reference-guides/data/data-core.md [hunk 2]
```diff
@@ -192,11 +192,11 @@ Returns the loaded entities for the given kind.
 _Parameters_
 
 -   _state_ `State`: Data state.
--   _kind_ `Kind`: Entity kind.
+-   _kind_ `ET.Kind`: Entity kind.
```
Comment (adamziel) [inline docs/reference-guides/data/data-core.md:195]: It's not ideal that the docs mention both ET.Kind and Kind, but tradeoffs, tradeoffs 🤷

File: packages/data/README.md [hunk 3]
```diff
@@ -174,11 +174,16 @@ const reduxStore = createStore();
 
 const boundSelectors = mapValues(
 	existingSelectors,
-	( selector ) => ( ...args ) => selector( reduxStore.getState(), ...args )
+	( selector ) =>
```
Comment (adamziel) [inline packages/data/README.md]: ESLint likes to format code in a different way now and it picked up README.md for some reason

**Comment by adamziel** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/42238

**Pull request closed without merging**